consensus/event_loop.go

```diff
@@ -0,0 +1,854 @@
+	// the server that we can construct a timeoutQC in the current view.
+	if timeoutQCEvent := fc.tryConstructTimeoutQCInCurrentView(); timeoutQCEvent != nil {
+		// Signal the server that we can construct a timeout QC for the current view, and mark
+		// that we have constructed a QC for the current view.
+		fc.hasConstructedQCInCurrentView = true
+		fc.Events <- timeoutQCEvent
+	}
+
+	return nil
+}
+
+// Sets the initial times for the crank timer and timeouts and starts scheduled tasks.
+func (fc *FastHotStuffEventLoop) Start() {
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Check if the event loop is either running or uninitialized.
+	// If it's running or uninitialized, then there's nothing to do here.
+	if fc.status != eventLoopStatusInitialized {
+		return
+	}
+
+	// Update the event loop's status to mark it as running.
+	fc.status = eventLoopStatusRunning
+
+	// Set the initial crank timer and timeout scheduled tasks
+	fc.resetScheduledTasks()
+}
+
+func (fc *FastHotStuffEventLoop) Stop() {
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Check if the event loop is no longer running. If it's not running
+	// we can simply return here.
+	if fc.status != eventLoopStatusRunning {
+		return
+	}
+
+	// Cancel the crank timer and timeout scheduled tasks, if any.
+	fc.crankTimerTask.Cancel()
+	fc.nextTimeoutTask.Cancel()
+
+	// Update the event loop's status so it is no longer marked as running.
+	fc.status = eventLoopStatusInitialized
+}
+
+func (fc *FastHotStuffEventLoop) IsInitialized() bool {
+	fc.lock.RLock()
+	defer fc.lock.RUnlock()
+
+	return fc.status != eventLoopStatusNotInitialized
+}
+
+func (fc *FastHotStuffEventLoop) IsRunning() bool {
+	fc.lock.RLock()
+	defer fc.lock.RUnlock()
+
+	return fc.status == eventLoopStatusRunning
+}
+
+// resetScheduledTasks recomputes the nextBlockConstructionTimeStamp and nextTimeoutTimeStamp
+// values, and reschedules the crank timer and timeout tasks.
+func (fc *FastHotStuffEventLoop) resetScheduledTasks() {
+	// Compute the next timeout ETA. We use exponential back-off for timeouts when there are
+	// multiple consecutive timeouts. We use the difference between the current view and the
+	// chain tip's view to determine this. The current view can only drift from the chain tip's
+	// view as a result of timeouts. This guarantees that the number of consecutive timeouts is
+	// always: max(currentView - tip.block.GetView() - 1, 0).
+
+	timeoutDuration := fc.timeoutBaseDuration
+
+	// Check if we have timed out at for the last n views. If so, we apply exponential
+	// back-off to the timeout base duration.
+	if fc.tip.block.GetView() < fc.currentView-1 {
+		// Note, there is no risk of underflow here because the following is guaranteed:
+		// currentView > tip.block.GetView() + 1.
+		numTimeouts := fc.currentView - fc.tip.block.GetView() - 1
+
+		// Compute the exponential back-off: nextTimeoutDuration * 2^numTimeouts
+		timeoutDuration = fc.timeoutBaseDuration << numTimeouts
```

Comment:
Nit: I generally prefer we avoid bitwise math where possible. It's inflexible, and often the source of errors because people miscalculated.

In the case of a duration, I'm not even 100% sure this is doing what we want since it's an int64. If there's a nice clean function you can use like Pow(2, exponent) or something, prefer that.